File: backend/rasero/dominio/identidad_cliente.py
"""Validación del identificador de un cliente: cédula ecuatoriana o RUC de persona natural
(User Story 4 de 002-clientes-fidelizacion, FR-017). Funciones puras: sin acceso a base de
datos ni a infraestructura, mismo estilo que `censura.py` y `fuga_cliente.py`.

El identificador es SIEMPRE opcional (FR-003, Principio II): sirve para no duplicar al mismo
cliente entre visitas cuando el cajero lo tiene a mano, nunca para bloquear una venta. Estas
funciones sólo responden "¿este texto es un identificador bien formado?"; la unicidad y el
rechazo 409 viven en el servicio/endpoint, no aquí.

Alcance deliberadamente acotado (un cliente individual de un minimarket):

  * Cédula: 10 dígitos. Los dos primeros son código de provincia. Se aceptan 01–24 (las 24
    provincias) y además el 30, asignado por el Registro Civil a ecuatorianos registrados en
    el exterior y a extranjeros residentes — negarlo excluiría a clientes reales con cédula
    válida. Tercer dígito 0–5 (persona natural). Décimo dígito verificador por módulo 10 con
    coeficientes 2,1,2,1,2,1,2,1,2 sobre los primeros 9 (cada producto > 9 se reduce restándole
    9; el verificador es el complemento a la decena de la suma módulo 10).
  * RUC de persona natural: 13 dígitos = una cédula válida (mismos 10 dígitos, mismo algoritmo)
    + sufijo literal "001".

NO se implementa el RUC de sociedad/empresa (tercer dígito 6 o 9, algoritmo módulo 11): no
aplica a un cliente individual. Cualquier texto que no calce con cédula ni con RUC natural es
inválido; no se intenta adivinar.
"""

_PROVINCIAS_VALIDAS = set(range(1, 25)) | {30}
_COEFICIENTES_CEDULA = (2, 1, 2, 1, 2, 1, 2, 1, 2)


def validar_cedula_ecuatoriana(numero: str) -> bool:
    """`True` si `numero` es una cédula ecuatoriana de persona natural bien formada."""
    if not isinstance(numero, str) or not numero.isdigit() or len(numero) != 10:
        return False

    provincia = int(numero[:2])
    if provincia not in _PROVINCIAS_VALIDAS:
        return False

    if int(numero[2]) > 5:  # tercer dígito 0–5: persona natural
        return False

    suma = 0
    for digito, coeficiente in zip(numero[:9], _COEFICIENTES_CEDULA):
        producto = int(digito) * coeficiente
        suma += producto - 9 if producto > 9 else producto

    verificador = (10 - suma % 10) % 10
    return verificador == int(numero[9])


def validar_identificador(numero: str) -> bool:
    """`True` si `numero` es una cédula (10 dígitos) o un RUC de persona natural (13 dígitos
    terminados en "001" cuyos primeros 10 son una cédula válida). Cualquier otro caso: `False`.
    """
    if not isinstance(numero, str) or not numero.isdigit():
        return False
    if len(numero) == 10:
        return validar_cedula_ecuatoriana(numero)
    if len(numero) == 13:
        return numero[10:] == "001" and validar_cedula_ecuatoriana(numero[:10])
    return False

File: frontend/src/dominio/identidadCliente.ts
/**
 * Réplica ligera en TypeScript de la validación de cédula / RUC de persona natural, para dar
 * feedback EN VIVO en el formulario antes de enviar (Parte 2).
 *
 * FUENTE DE VERDAD DEL ALGORITMO: `backend/rasero/dominio/identidad_cliente.py`. El backend
 * sigue siendo el que decide (422 `identificador_invalido`); esto es solo una ayuda de UI y
 * debe mantenerse igual a esa función. Cualquier cambio del algoritmo se hace allí primero.
 *
 * - Cédula: 10 dígitos. Provincia (2 primeros) en 01–24 o 30. Tercer dígito 0–5 (persona
 *   natural). Décimo dígito verificador por módulo 10 (coef. 2,1,2,1,2,1,2,1,2; producto > 9
 *   se reduce restándole 9; verificador = complemento a la decena de la suma módulo 10).
 * - RUC de persona natural: 13 dígitos = cédula válida + sufijo "001".
 * - RUC de sociedad (módulo 11): fuera de alcance.
 */

const PROVINCIAS_VALIDAS = new Set<number>([...Array.from({ length: 24 }, (_, i) => i + 1), 30]);
const COEFICIENTES = [2, 1, 2, 1, 2, 1, 2, 1, 2];

export function validarCedulaEcuatoriana(numero: string): boolean {
  if (!/^\d{10}$/.test(numero)) return false;

  const provincia = Number(numero.slice(0, 2));
  if (!PROVINCIAS_VALIDAS.has(provincia)) return false;

  if (Number(numero[2]) > 5) return false;

  let suma = 0;
  for (let i = 0; i < 9; i++) {
    const producto = Number(numero[i]) * COEFICIENTES[i];
    suma += producto > 9 ? producto - 9 : producto;
  }
  const verificador = (10 - (suma % 10)) % 10;
  return verificador === Number(numero[9]);
}

export function validarIdentificador(numero: string): boolean {
  if (!/^\d+$/.test(numero)) return false;
  if (numero.length === 10) return validarCedulaEcuatoriana(numero);
  if (numero.length === 13) return numero.slice(10) === "001" && validarCedulaEcuatoriana(numero.slice(0, 10));
  return false;
}
